File: python/data-structures/point.py
from math import sqrt

# point(number, number) -> tuple
# Constructs a two dimensional point as a tuple.
# point(1, 2) -> (1, 2)

def point(x, y):
    return (x, y)

# distance(tuple, tuple) -> float
# Calculates the distance between two points.
# distance((3, 0), (2, 0)) -> 1.0

def distance(p1, p2):
    def hypotenuse(x, y):
        return sqrt((x * x) + (y * y))
    x1, y1 = p1 
    x2, y2 = p2
    return hypotenuse(x2 - x1, y2 - y1)

# path(points) -> tuple
# Constructs a path as a tuple of points.
# path(point(1, 2), point(3, 4)) -> ((1, 2), (3, 4))

def path(*points):
    return points

File: javascript/data-structures/point.js
// hypotenuse(...numbers) -> number
// Returns the square root of the sum of squares of its arguments.
// hypotenuse(3, 4) -> 5

function hypotenuse(...numbers) {
    return Math.hypot(...numbers);
}

// point(number, number) -> array
// Constructs a two-dimensional point represented as an immutable array.
// point(1, 2) -> [1, 2]

function point(x, y) {
    return Object.freeze([x, y]);
}

// ORIGIN: point of origin in a two-dimensional coordinate system.

const ORIGIN = point(0, 0);

// segment(array, array) -> number
// Calculates the distance between two points.
// segment(point(3, 0), point(2, 0)) -> 1

function segment(p1, p2) {
    const [x1, y1] = p1;
    const [x2, y2] = p2;
    return hypotenuse(x1 - x2, y1 - y2);
}

// path(...point) -> array
// Constructs an array of two-dimensional points.
// path(point(1, 2), point(3, 4)) -> [[1, 2], [3, 4]]

function path(...points) {
    return Object.freeze(points);
}
